mango-python/bdgenomics/mango/variantViz.py

```diff
@@ -0,0 +1,443 @@
+#
+
+from collections import Counter, OrderedDict
+from cigar import Cigar
+import matplotlib.pyplot as plt
+import numpy as np
+import matplotlib.ticker as ticker
+import math
+from pyspark.sql import Row
+plt.rcdefaults()
+plt.rcParams['figure.figsize'] = [10, 8]
+
+default_bins=100
+
+
+def findMostSigEffect(effectList):
+    """
+    Find and return the most deleterious effect from a list of Sequence Ontology terms
+    :param effectList: list of sequence ontology (so) terms
+    :return: so_term from effectList with most deleterious effect
+    """
+
+    # so_terms, listed in order of least to most clinically significant (deleterious), for purpose of selecting the
+    # most important effect of a given variant.
+    so_terms = [
+     'intergenic_region',
+     'downstream_gene_variant',
+      'upstream_gene_variant',
+     'sequence_feature',
+     'structural_interaction_variant',
+     'intron_variant',
+     'non_coding_transcript_exon_variant',
+     '3_prime_UTR_variant',
+     '5_prime_UTR_variant',
+     'synonymous_variant',
+     'conservative_inframe_insertion',
+     'conservative_inframe_deletion',
+     'initiator_codon_variant',
+     'protein_protein_contact',
+     'non_coding_transcript_variant',
+     'stop_retained_variant',
+     '5_prime_UTR_premature_start_codon_gain_variant',
+     'disruptive_inframe_deletion',
+     'disruptive_inframe_insertion',
+     'missense_variant',
+     'stop_lost',
+     'start_lost',
+     'splice_donor_variant',
+     'splice_acceptor_variant',
+     'splice_region_variant',
+     'frameshift_variant',
+     'stop_gained']
+    so_to_rank = dict(zip(so_terms, range(1,len(so_terms)+1)))
+    rank_to_so = {v: k for k, v in so_to_rank.iteritems()}
+    return rank_to_so[max( [ so_to_rank[item] for item in effectList] )]
+
+class VariantFreqPopCompareScatter(object):
+    """
+    Scattergram comparing allele frequencies of variants in two populations
+    """
+    def __init__(self, ss, variantRDD, pop1='AF_NFE', pop2='AF_AFR'):
+        """
+        :param ss: SparkContext
+        :param variantRDD: bdgenomics variantRDD
+        :param pop1: population one name, default AF_NFE
+        :param pop2: population two name, default AF_AFR
+        """
+        curr = variantRDD.toDF().rdd.map(lambda w: w.asDict()).map(lambda p: p['annotation']).filter(
+            lambda x: x['attributes'][pop1] != ".").filter(
+            lambda x: x['attributes'][pop2] != ".").map(lambda d: ((float(d['attributes'][pop1])), float(d['attributes'][pop2]))   ).collect()
+        self.x = [ i[0] for i in curr ]
+        self.y = [ i[1] for i in curr ]
+        self.pop1 = pop1
+        self.pop2 = pop2
+
+    def plot(self):
+        plt.rcdefaults()
+        plt.title("Population Frequency\ncomparing populations " + self.pop1 + " and " + self.pop2)
+        mag = [ math.log(max( ( (z[0]-z[1]/(z[0]+.000000000001)), (z[1]-z[0])/(z[1]+.000000000001)))+1) for z in zip(self.x,self.y) ]
```

Comment:
The .000000000001 is just a hack to avoid a division by zero when calculating a value used to color the points in graph calculated by the max ( log( x-y/ y , log(y-x/x) ) which seemed to be a visually effective way to highlight points with a more extreme difference in frequency between the two populations.

Reply:
Hard coding numbers should be avoided. If absolutely necessary, they should be well documented

Reply:
OK - I'll add a comment documenting them.